include/nil/crypto3/zk/commitments/polynomial/kzg_v2.hpp

```diff
@@ -0,0 +1,348 @@
+                    void mark_batch_as_fixed(std::size_t index) {
+                    }
+
+                    static params_type create_params(std::size_t d, typename KZGScheme::scalar_value_type alpha) {
+                        return params_type(d, 1, alpha);
+                    }
+
+                    kzg_commitment_scheme_v2(params_type kzg_params) : _params(kzg_params) {
+                        BOOST_ASSERT( kzg_params.verification_key.size() == 2);
+                    }
+
+                    // Differs from static, because we pack the result into byte blob.
+                    commitment_type commit(std::size_t index){
+                        this->_ind_commitments[index] = {};
+                        this->state_commited(index);
+
+                        std::vector<std::uint8_t> result = {};
+                        for (std::size_t i = 0; i < this->_polys[index].size(); ++i) {
+                            BOOST_ASSERT(this->_polys[index][i].degree() <= _params.commitment_key.size());
+                            auto single_commitment = nil::crypto3::zk::algorithms::commit_one<KZGScheme>(_params, this->_polys[index][i]);
+                            this->_ind_commitments[index].push_back(single_commitment);
+                            nil::marshalling::status_type status;
+                            std::vector<uint8_t> single_commitment_bytes =
+                                nil::marshalling::pack<endianness>(single_commitment, status);
+                            BOOST_ASSERT(status == nil::marshalling::status_type::success);
+                            result.insert(result.end(), single_commitment_bytes.begin(), single_commitment_bytes.end());
+                        }
+                        _commitments[index] = result;
+                        return result;
+                    }
+
+                    using preprocessed_data_type = bool;
+                    preprocessed_data_type preprocess(transcript_type& transcript) const{
+                        return true;
+                    }
+
+                    void setup(transcript_type& transcript, preprocessed_data_type b = true) {
+                        // Nothing to be done here.
+                    }
+
+                    proof_type proof_eval(transcript_type &transcript){
+                        this->eval_polys();
+                        this->merge_eval_points();
+
+                        for( auto const &it: this->_commitments ){
+                            auto k = it.first;
+                            update_transcript(k, transcript);
+                        }
+
+                        auto theta = transcript.template challenge<typename KZGScheme::curve_type::scalar_field_type>();
+                        auto theta_i = KZGScheme::scalar_value_type::one();
+                        auto f = math::polynomial<typename KZGScheme::scalar_value_type>::zero();
+
+                        for( auto const &it: this->_polys ){
```

Comment:
This one also need to be parallelized, it uses "this->_polys[k][i].coefficients()", which does FFT, which is slow. You will need to use the HIGH level thread pool here.